File: packages/server/src/tool-router.ts
/**
 * @fileoverview Shared MCP tool call routing — dispatches tool calls to the
 * appropriate execution function (batch, pipeline, call, lazy discovery, grouped, or individual).
 *
 * Used by both stdio and HTTP transports to eliminate routing duplication.
 */

import type {
	BatchRequest,
	BatchResult,
	CommandContext,
	CommandResult,
	PipelineRequest,
	PipelineResult,
} from '@lushly-dev/afd-core';
import { failure, isBatchRequest, isPipelineRequest, truncateName } from '@lushly-dev/afd-core';
import type { ContextState } from './bootstrap/afd-context.js';
import {
	commandAction,
	commandGroup,
	filterByContext,
	type GroupByFn,
	isAccessibleInContext,
	notFoundSuggestion,
	notInContextError,
} from './command-routing.js';
import { resolveContextState } from './context-scope.js';
import { executeDetail, executeDiscover } from './lazy-tools.js';
import {
	batchArgsSchema,
	callArgsSchema,
	checkMetaArgs,
	detailArgsSchema,
	discoverArgsSchema,
	parseMetaArgs,
	pipeArgsSchema,
} from './meta-tools.js';
import type { ZodCommandDefinition } from './schema.js';

// ═══════════════════════════════════════════════════════════════════════════════
// TYPES
// ═══════════════════════════════════════════════════════════════════════════════

export interface ToolCallResult {
	content: Array<{ type: string; text: string }>;
	isError: boolean;
}

export interface ToolRouterDeps {
	executeCommand: (
		name: string,
		input: unknown,
		context?: CommandContext
	) => Promise<CommandResult>;
	executeBatch: (request: BatchRequest, context?: CommandContext) => Promise<BatchResult>;
	executePipeline: (request: PipelineRequest, context?: CommandContext) => Promise<PipelineResult>;
	commands: ZodCommandDefinition[];
	toolStrategy: 'individual' | 'grouped' | 'lazy';
	groupByFn?: GroupByFn;
	devMode: boolean;
	/** All registered commands, including those outside the commands array */
	allCommands?: ZodCommandDefinition[];
	/** Set of command names exposed via the server's commands array */
	exposedCommandNames?: Set<string>;
	/** Context state used when the call context carries none (stdio and tests). */
	contextState?: ContextState;
}

// ═══════════════════════════════════════════════════════════════════════════════
// EMPTY PIPELINE RESULT (for invalid pipeline requests)
// ═══════════════════════════════════════════════════════════════════════════════

const emptyPipelineResult = {
	data: undefined,
	metadata: {
		confidence: 0,
		confidenceBreakdown: [],
		reasoning: [],
		warnings: [],
		sources: [],
		alternatives: [],
		executionTimeMs: 0,
		completedSteps: 0,
		totalSteps: 0,
	},
	steps: [],
};

// ═══════════════════════════════════════════════════════════════════════════════
// ROUTER
// ═══════════════════════════════════════════════════════════════════════════════

/**
 * Wrap a result as MCP tool content. The JSON is compact: the text is read by an
 * agent, and indentation only adds tokens.
 */
function resultContent(data: unknown, isError: boolean): ToolCallResult {
	return {
		content: [{ type: 'text', text: JSON.stringify(data) }],
		isError,
	};
}

/**
 * The `COMMAND_NOT_IN_CONTEXT` tool result when `cmd` exists but is outside the
 * active context; undefined otherwise (a missing command is left to execution).
 */
function contextDenial(
	cmd: ZodCommandDefinition | undefined,
	commandName: string,
	contextState: ContextState | null | undefined
): ToolCallResult | undefined {
	const activeContext = contextState?.getActive();
	if (!cmd || !activeContext || isAccessibleInContext(cmd, activeContext)) return undefined;
	return resultContent(
		{ success: false, error: notInContextError(commandName, activeContext) },
		true
	);
}

/**
 * Create a tool call router that dispatches to the appropriate execution function.
 */
export function createToolRouter(deps: ToolRouterDeps) {
	const {
		executeCommand,
		executeBatch,
		executePipeline,
		commands,
		toolStrategy,
		groupByFn,
		devMode,
		allCommands,
		exposedCommandNames,
	} = deps;

	/**
	 * Route one tool call.
	 *
	 * @param callContext - Base `CommandContext` for commands this call runs (request values,
	 *   `signal`, and the caller's context state); each execution adds its own `traceId`.
	 */
	return async function routeToolCall(
		toolName: string,
		args: unknown,
		callContext: CommandContext = {}
	): Promise<ToolCallResult> {
		const contextState = resolveContextState(callContext, deps.contextState);
		const traced = (prefix: string): CommandContext => ({
			...callContext,
			traceId: `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2)}`,
		});

		// Handle afd-call (available in all strategies)
		if (toolName === 'afd-call') {
			const parsed = parseMetaArgs(
				toolName,
				callArgsSchema,
				args,
				'Provide { command: "command-name", input: {...} }'
			);
			if (!parsed.success) return resultContent(parsed.result, true);
			const { command: commandName, input } = parsed.data;

			// Check if command exists in the exposed commands
			const cmd = commands.find((c) => c.name === commandName);
			if (!cmd) {
				// Check if it exists but isn't exposed
				const allCmd = allCommands?.find((c) => c.name === commandName);
				if (allCmd) {
					return resultContent(
						{
							success: false,
							error: {
								code: 'COMMAND_NOT_EXPOSED',
								message: `Command '${commandName}' exists but is not exposed via this server`,
								suggestion:
									"This command is registered but not included in the server's commands array.",
							},
						},
						true
					);
				}
				// Not found at all — suggest a few close matches the caller can use
				const callable = filterByContext(commands, contextState?.getActive());
				return resultContent(
					{
						success: false,
						error: {
							code: 'COMMAND_NOT_FOUND',
							message: `Command '${truncateName(commandName)}' not found`,
							suggestion: notFoundSuggestion(
								commandName,
								callable.map((c) => c.name)
							),
						},
					},
					true
				);
			}

			// Context check: validate the command is accessible in current context
			const denied = contextDenial(cmd, commandName, contextState);
			if (denied) return denied;

			// Execute through the execution engine (full middleware chain)
			const result = await executeCommand(commandName, input ?? {}, traced('afd-call'));
			return resultContent(result, !result.success);
		}

		// Handle afd-discover (listed by the lazy strategy, routable in all)
		if (toolName === 'afd-discover') {
			const parsed = parseMetaArgs(
				toolName,
				discoverArgsSchema,
				args,
				'Provide optional { category, tag, tagMode, search, includeMutation, limit, offset }'
			);
			if (!parsed.success) return resultContent(parsed.result, true);
			// Filter by active context when context state is available
			const discoverCommands = filterByContext(commands, contextState?.getActive());
			const result = executeDiscover(discoverCommands, parsed.data);
			return resultContent(result, false);
		}

		// Handle afd-detail (listed by the lazy and grouped strategies, routable in all)
		if (toolName === 'afd-detail') {
			const parsed = parseMetaArgs(
				toolName,
				detailArgsSchema,
				args,
				'Provide { command: "command-name" } or { command: ["name-1", "name-2"] } (max 10)'
			);
			if (!parsed.success) return resultContent(parsed.result, true);
			const exposedNames = exposedCommandNames ?? new Set(commands.map((c) => c.name));
			// Filter by active context so agents cannot inspect commands outside their context
			const contextFilteredCmds = filterByContext(
				allCommands ?? commands,
				contextState?.getActive()
			);
			const result = executeDetail(contextFilteredCmds, exposedNames, parsed.data);
			return resultContent(result, false);
		}

		// Handle built-in afd-batch tool
		if (toolName === 'afd-batch') {
			// The Zod schema pinpoints shape errors; isBatchRequest is the envelope check
			// shared with POST /batch. Both report INVALID_BATCH_REQUEST.
			const check = checkMetaArgs(batchArgsSchema, args);
			if (!check.success || !isBatchRequest(args)) {
				const usage = 'Provide { commands: [...] } with command objects';
				return resultContent(
					{
						success: false,
						error: {
							code: 'INVALID_BATCH_REQUEST',
							message: 'Invalid batch request format',
							suggestion: check.success ? usage : `${check.suggestion}. ${usage}`,
							...(!check.success && { details: { errors: check.errors } }),
						},
					},
					true
				);
			}
			const result = await executeBatch(args, traced('batch'));
			return resultContent(result, !result.success);
		}

		// Handle built-in afd-pipe tool
		if (toolName === 'afd-pipe') {
			// As for afd-batch: Zod pinpoints shape errors, and isPipelineRequest checks the
			// whole envelope (including `when` conditions). Both report INVALID_PIPELINE_REQUEST.
			const check = checkMetaArgs(pipeArgsSchema, args);
			if (!check.success || !isPipelineRequest(args)) {
				const usage =
					'Provide steps with nonempty command names, object inputs, valid conditions, and correctly typed options; input must be JSON';
				return resultContent(
					{
						...emptyPipelineResult,
						steps: [
							{
								index: -1,
								command: '',
								status: 'failure',
								executionTimeMs: 0,
								error: {
									code: 'INVALID_PIPELINE_REQUEST',
									message: 'Invalid pipeline request envelope',
									suggestion: check.success ? usage : `${check.suggestion}. ${usage}`,
									...(!check.success && { details: { errors: check.errors } }),
								},
							},
						],
					},
					true
				);
			}
			const result = await executePipeline(args, traced('pipeline'));
			const hasFailed = result.steps.some((s) => s.status === 'failure');
			return resultContent(result, hasFailed);
		}

		// Handle grouped tool calls (when toolStrategy === "grouped")
		if (toolStrategy === 'grouped') {
			const typedArgs = args as { action?: string; params?: unknown } | undefined;
			const action = typedArgs?.action;
			const commandParams = typedArgs?.params ?? {};

			if (devMode) {
				console.error(
					`[MCP Debug] Grouped tool call: toolName=${toolName}, action=${action}, args=${JSON.stringify(args)}`
				);
			}

			if (action && typeof action === 'string') {
				const candidates = commands.filter(
					(command) =>
						commandGroup(command, groupByFn) === toolName && commandAction(command) === action
				);
				if (candidates.length !== 1) {
					return resultContent(
						failure({
							code: candidates.length ? 'AMBIGUOUS_ACTION' : 'COMMAND_NOT_FOUND',
							message: `Action '${truncateName(action)}' does not identify one command in group '${truncateName(toolName)}'`,
							suggestion:
								'Use afd-call with the full command name, or list tools to find a valid action.',
						}),
						true
					);
				}
				const actualCommandName = candidates[0]?.name as string;

				// Context check for grouped strategy
				const denied = contextDenial(candidates[0], actualCommandName, contextState);
				if (denied) return denied;

				const result = await executeCommand(actualCommandName, commandParams, traced('trace'));
				return resultContent(result, !result.success);
			}

			// In grouped mode, if tool is a known group but action is missing/invalid,
			// return a helpful error instead of falling through
			const groupCommands = commands.filter((cmd) => commandGroup(cmd, groupByFn) === toolName);

			if (groupCommands.length > 0) {
				const availableActions = groupCommands.map(commandAction);

				return resultContent(
					{
						success: false,
						error: {
							code: 'INVALID_GROUPED_CALL',
							message: `Grouped tool '${toolName}' requires an 'action' parameter`,
							suggestion: `Provide { action: "<action>", params: {...} }. Available actions: ${availableActions.join(', ')}`,
						},
						_debug: devMode ? { receivedArgs: args } : undefined,
					},
					true
				);
			}
		}

		// Handle user-defined commands (individual mode or direct command calls)
		// Context check: validate the command is accessible in current context
		const denied = contextDenial(
			commands.find((c) => c.name === toolName),
			toolName,
			contextState
		);
		if (denied) return denied;

		const result = await executeCommand(toolName, args ?? {}, traced('trace'));
		return resultContent(result, !result.success);
	};
}

File: python/src/afd/server/tool_router.py
"""Shared MCP tool call routing for AFD servers."""

from __future__ import annotations

import time
import uuid
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any

from afd.core.batch import BatchRequest, is_batch_request
from afd.core.commands import CommandContext, CommandDefinition
from afd.core.pipeline import PipelineRequest, is_pipeline_request
from afd.core.result import CommandResult, error
from afd.server.lazy_tools import execute_detail, execute_discover
from afd.server.tools import derive_group_action, derive_group_name


@dataclass
class ToolRouterDeps:
    execute_command: Callable[[str, Any, CommandContext | None], Awaitable[CommandResult[Any]]]
    execute_batch: Callable[[BatchRequest | dict[str, Any], CommandContext | None], Awaitable[Any]]
    execute_pipeline: Callable[[PipelineRequest | dict[str, Any], CommandContext | None], Awaitable[Any]]
    commands: list[CommandDefinition]
    tool_strategy: str
    group_by_fn: Callable[[CommandDefinition], str | None] | None = None
    all_commands: list[CommandDefinition] | None = None
    exposed_command_names: set[str] | None = None
    context_state: Any = None


def new_trace_id(prefix: str) -> str:
    """A unique trace ID for one tool call, like the TypeScript server's."""
    return f"{prefix}-{int(time.time() * 1000)}-{uuid.uuid4().hex[:10]}"


def is_command_accessible(
    command: CommandDefinition | None,
    active_context: str | None,
) -> bool:
    """Return whether the command is visible in the active context."""
    if not active_context or command is None:
        return True
    if not command.contexts:
        return True
    return active_context in command.contexts


def _index_by_name(commands: list[CommandDefinition]) -> dict[str, CommandDefinition]:
    """Name -> command; the first command with a name wins, as a linear search would."""
    index: dict[str, CommandDefinition] = {}
    for command in commands:
        index.setdefault(command.name, command)
    return index


def create_tool_router(deps: ToolRouterDeps):
    """Create a router that dispatches built-in, grouped, or direct MCP tools.

    Lookups are precomputed when the router is created, so a call costs the
    same at 10 or 1,000 commands. The active context is read on every call.
    """
    all_commands = deps.all_commands or deps.commands
    exposed_command_names = deps.exposed_command_names or {
        command.name for command in deps.commands
    }
    commands_by_name = _index_by_name(deps.commands)
    all_commands_by_name = _index_by_name(all_commands)
    get_group = deps.group_by_fn or derive_group_name
    commands_by_group: dict[str, list[CommandDefinition]] = {}
    all_commands_by_group: dict[str, list[CommandDefinition]] = {}
    if deps.tool_strategy == "grouped":
        for command in deps.commands:
            commands_by_group.setdefault(get_group(command) or "general", []).append(command)
        for command in all_commands:
            all_commands_by_group.setdefault(get_group(command) or "general", []).append(command)

    def mcp_context(prefix: str, active_context: str | None) -> CommandContext:
        return CommandContext(
            trace_id=new_trace_id(prefix),
            extra={"interface": "mcp", "active_context": active_context},
        )

    def not_in_context(name: str, active_context: str | None) -> CommandResult[Any]:
        return error(
            "COMMAND_NOT_IN_CONTEXT",
            f"Command '{name}' is not available in context '{active_context}'",
            suggestion="Use afd-context-list to see contexts or afd-context-enter to switch.",
        )

    async def route_tool_call(tool_name: str, args: Any) -> Any:
        active_context = deps.context_state.get_active() if deps.context_state else None

        if tool_name == "afd-call":
            payload = args or {}
            command_name = payload.get("command") if isinstance(payload, dict) else None
            if not command_name or not isinstance(command_name, str):
                return error(
                    "VALIDATION_ERROR",
                    "Missing required field: command",
                    suggestion="Provide { command: 'command-name', input: {...} }.",
                )

            command = commands_by_name.get(command_name)
            if command is None:
                registered = all_commands_by_name.get(command_name)
                if registered is not None:
                    if not is_command_accessible(registered, active_context):
                        return not_in_context(command_name, active_context)
                    return error(
                        "COMMAND_NOT_EXPOSED",
                        f"Command '{command_name}' exists but is not exposed via this server",
                        suggestion="Enable MCP exposure for the command or use a different interface.",
                    )
                return error(
                    "COMMAND_NOT_FOUND",
                    f"Command '{command_name[:128]}' not found",
                    suggestion="Use afd-discover or afd-help to list available commands.",
                )

            if not is_command_accessible(command, active_context):
                return not_in_context(command_name, active_context)

            return await deps.execute_command(
                command_name,
                payload.get("input", {}),
                mcp_context("afd-call", active_context),
            )

        if tool_name == "afd-discover":
            visible_commands = [
                command
                for command in deps.commands
                if is_command_accessible(command, active_context)
            ]
            return execute_discover(visible_commands, args or {})

        if tool_name == "afd-detail":
            visible_all_commands = [
                command
                for command in all_commands
                if is_command_accessible(command, active_context)
            ]
            return execute_detail(visible_all_commands, exposed_command_names, args or {})

        if tool_name == "afd-batch":
            if not is_batch_request(args):
                return error(
                    "INVALID_BATCH_REQUEST",
                    "Invalid batch request format",
                    suggestion="Provide { commands: [...] } with command objects.",
                )
            return await deps.execute_batch(args, mcp_context("batch", active_context))

        if tool_name == "afd-pipe":
            if not is_pipeline_request(args):
                return error(
                    "INVALID_PIPELINE_REQUEST",
                    "Invalid pipeline request format",
                    suggestion="Provide { steps: [...] } with pipeline step objects.",
                )
            return await deps.execute_pipeline(args, mcp_context("pipeline", active_context))

        if deps.tool_strategy == "grouped":
            payload = args or {}
            action = payload.get("action") if isinstance(payload, dict) else None
            all_group_commands = all_commands_by_group.get(tool_name, [])
            group_commands = [
                command
                for command in commands_by_group.get(tool_name, [])
                if is_command_accessible(command, active_context)
            ]

            if not group_commands and all_group_commands:
                if isinstance(action, str):
                    hidden_command = next(
                        (
                            item
                            for item in all_group_commands
                            if derive_group_action(item) == action
                            and not is_command_accessible(item, active_context)
                        ),
                        None,
                    )
                    if hidden_command is not None:
                        return not_in_context(hidden_command.name, active_context)

                inaccessible_commands = [
                    item for item in all_group_commands if not is_command_accessible(item, active_context)
                ]
                if inaccessible_commands:
                    return error(
                        "COMMAND_NOT_IN_CONTEXT",
                        f"Grouped tool '{tool_name}' is not available in context '{active_context}'",
                        suggestion="Use afd-context-list to see contexts or afd-context-enter to switch.",
                    )

            if group_commands and not isinstance(action, str):
                available_actions = [derive_group_action(command) for command in group_commands]
                return error(
                    "INVALID_GROUPED_CALL",
                    f"Grouped tool '{tool_name}' requires an action parameter",
                    suggestion=(
                        "Provide { action: '<action>', params: {...} }. "
                        f"Available actions: {', '.join(available_actions)}"
                    ),
                )

            if group_commands and isinstance(action, str):
                command = next(
                    (
                        item
                        for item in group_commands
                        if derive_group_action(item) == action
                    ),
                    None,
                )
                if command is None:
                    hidden_command = next(
                        (
                            item
                            for item in all_group_commands
                            if derive_group_action(item) == action
                            and not is_command_accessible(item, active_context)
                        ),
                        None,
                    )
                    if hidden_command is not None:
                        return not_in_context(hidden_command.name, active_context)
                    return error(
                        "COMMAND_NOT_FOUND",
                        f"Action '{action}' is not available for grouped tool '{tool_name}'",
                        suggestion="Use afd-discover or afd-help to inspect available actions.",
                    )
                return await deps.execute_command(
                    command.name,
                    payload.get("params", {}),
                    mcp_context("trace", active_context),
                )

        command = commands_by_name.get(tool_name)
        if not is_command_accessible(command, active_context):
            return not_in_context(tool_name, active_context)

        return await deps.execute_command(
            tool_name,
            args or {},
            mcp_context("trace", active_context),
        )

    return route_tool_call
